File: First-Edition/BookSamples/CH08/src/Cellular/Language/Automaton.hs
{-# LANGUAGE DataKinds      #-}
{-# LANGUAGE GADTs          #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TypeFamilies   #-}
module Cellular.Language.Automaton where

import GHC.TypeLits
import Data.Proxy ( Proxy(..) )
import Data.Maybe (fromMaybe)
import qualified Data.Map as Map
import Control.Monad (mapM)


-- Interfaces

-- data IBoard where
--   BoardWrapper :: a -> IBoard

-- type family MkBoard a :: IBoard where
--   MkBoard a = BoardWrapper a

data ICellCondition where
  CellConditionWrapper :: a -> ICellCondition

type family MkCellCondition a :: ICellCondition where
  MkCellCondition a = CellConditionWrapper a

data IState where
  StateWrapper :: a -> IState

type family MkState a :: IState where
  MkState a = StateWrapper a

data IRule where
  RuleWrapper :: a -> IRule

type family MkRule a :: IRule where
  MkRule a = RuleWrapper a


-- Implementations

data NeighborsCountImpl
  (state :: IState)
  (counts :: [Nat])
type NeighborsCount s c =
  MkCellCondition (NeighborsCountImpl s c)

data CustomStateTransition = StateTransition
  { cstFromState :: IState
  , cstToState   :: IState
  , cstCondition :: ICellCondition
  }

data CustomStep = Step
  { defState :: IState
  , transitions :: [CustomStateTransition]
  }

data StateImpl
  (name :: Symbol)
type State n = MkState (StateImpl n)

data RuleImpl
  (name :: Symbol)
  (code :: Symbol)
  (step :: CustomStep)
type Rule n c s = MkRule (RuleImpl n c s)


type D = State "Dead"
type A = State "Alive"

type Neighbors3  = NeighborsCount A '[3  ]
type Neighbors23 = NeighborsCount A '[2,3]

type GoLStep = Step D
  '[ 'StateTransition D A Neighbors3
   , 'StateTransition A A Neighbors23
   ]

type GoLRule = Rule
  "Game of Life"
  "gol"
  GoLStep

File: First-Edition/BookSamples/rosetta-stone/scala3/rsc2p4_correctness_validation/src/main/scala/cellular/language/Automaton.scala
package cellular
package language

import graninas.typelevel._

// Interfaces

sealed trait IState extends IInterface
type MkState[A] = IState {
  type Impl = A
}

sealed trait ICellCondition extends IInterface
type MkCellCondition[A] = ICellCondition {
  type Impl = A
}

sealed trait INeighborhood extends IInterface
type MkNeighborhood[A] = INeighborhood {
  type Impl = A
}

sealed trait IStateTransition extends IInterface
type MkStateTransition[A] = IStateTransition {
  type Impl = A
}

sealed trait IStep extends IInterface
type MkStep[A] = IStep {
  type Impl = A
}

sealed trait IRule extends IInterface
type MkRule[A] = IRule {
  type Impl = A
}
